client/files.go

```diff
@@ -0,0 +1,195 @@
+// Copyright (c) 2022 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package client
+
+import (
+	"errors"
+	"io/fs"
+	"net/url"
+	"os"
+	"strconv"
+	"time"
+)
+
+type ListFilesOptions struct {
+	Path    string
+	Pattern string // optional
+	Itself  bool   // optional
+}
+
+type fileInfo struct {
+	name            string
+	size            int64
+	mode            fs.FileMode
+	modTime         time.Time
+	path            string
+	userID, groupID int
+	user, group     string
+}
+
+func (fi fileInfo) Name() string {
+	return fi.name
+}
+
+func (fi fileInfo) Size() int64 {
+	return fi.size
+}
+
+func (fi fileInfo) Mode() fs.FileMode {
+	return fi.mode
+}
+
+func (fi fileInfo) ModTime() time.Time {
+	return fi.modTime
+}
+
+func (fi fileInfo) IsDir() bool {
+	return fi.mode.IsDir()
+}
+
+func (fi fileInfo) Sys() interface{} {
+	return nil
+}
+
+func (fi fileInfo) Path() string {
+	return fi.path
+}
+
+func (fi fileInfo) UserID() int {
+	return fi.userID
+}
+
+func (fi fileInfo) GroupID() int {
+	return fi.groupID
+}
+
+func (fi fileInfo) User() string {
+	return fi.user
+}
+
+func (fi fileInfo) Group() string {
+	return fi.group
+}
+
+type FileInfo interface {
+	os.FileInfo
```

Comment:
About option 1, I think it would be a bit counter-intuitive to have it return a `fs.FileInfo` which can be type asserted to `client.FileInfo`. How does the user of our API know this information? Also, since the `client.FileInfo` struct would implement `fs.FileInfo`, it would be usable with functions where `fs.FileInfo` is accepted anyway.

I like the option 2 better which, as you say, is more consistent with the rest of the API. Regarding the struct implementing `fs.FileInfo`, would it look like the following?

```go
type FileInfo struct {
	fs.FileInfo

	Path            string
	UserID, GroupID int
	User, Group     string
}
```

I'm asking because the contrast between having `FileInfo.Name()`, `FileInfo.Mode()`, et al. but then `FileInfo.UserID` and not `FileInfo.UserID()` looks a bit inconsistent to me—but if it's considered more idiomatic I'd go with that (my Go is still a bit bad, also 😅).

Thank you very much for your feedback!

Reply:
Yeah, I'd probably make them all methods, even `Path()`, `User()`, and so on that aren't part of the `fs.FileInfo` interface, just for consistency. It does seem weird to have some of them methods and some fields.